contributors/design-proposals/code-generator/kubegen.md

```diff
@@ -0,0 +1,333 @@
+This will use the contents from the boilerplate.go.txt file instead of LICENSE.
+If boilerplate.go.txt is already in comments then use it verbatim, otherwise wrap it in comments.
+
+
+**Note**: This passes the `--go-header-file` to each of the code generators
+
+```sh
+kubegen versioned --license-file "boilerplate.go.txt"
+```
+
+#### Define the license type
+
+If no license file exists, one may be specified on the using the `--license` flag.
+
+```sh
+kubegen versioned --license "Apache 2.0"
+```
+
+Options:
+
+- "Apache 2.0"
+- "None"
+
+#### Define API groups and versions
+
+The set of APIs code generation is performed for can be explicitly defined instead of
+implicitly determined using the `--api-version` and `--api-group` flags.  When these
+flags are provided, API groups will not be implicitly discovered.  Only API versions
+for groups specified with `--api-group` will be discovered.
+
+```sh
+kubegen versioned --api-version apps/v1 --api-version apps/v1beta1 --api-version extensions/v1beta1
+```
+
+#### Manually specifying the output for the versioned client packages
+
+By default versioned client related libraries are output to the following locations:
+
+- client: `pkg/client/clientset_generated/clientset`
+- informers: `pkg/client/informers_generated/externalversions`
+- listers: `pkg/client/listers_generated`
+
+These locations can be changed using the flags `--client-output`, ``--informer-output` and `--listers-output`.
+This is useful if generating a client for external consumption.
+
+**Note**: These flags also work with `kubegen internal` since that runs generators on versioned types.
+
+```sh
+kubegen versioned --client-output kubernetes --informers-output informers --listers-output listers
+```
+
+#### Running in dry-run
+
+To verify the output without writing actual files use the `--dry-run` flag
+
+```sh
+kubegen versioned --dry-run
+```
+
+**Note**: This passes the `--verify-only` flag to each of the code generators
+
+  
+#### Overriding the searched the api directories
+
+The `--apis-dir` defaults to `pkg/apis` and looks for API groups in that directory.
+
+Looking for APIs outside the default location can be configured using
+the `--apis-dir` flag.  This flag may be provided multiple times to search multiple
+directories.
+
+```sh
+kubegen versioned --apis-dir notpkg/apis --apis-dir pkg/notapis
+```
+
+## Running via Bazel
+
+### Installing via Bazel
+
+Add the following to the project *WORKSPACE* file
+
+```py
+http_archive(
+    name = "io_k8s_rules_go",
+    url = "https://github.com/kubernetes/bazelbuild/releases/download/v1.8.0/rules_go-1.8.0.tar.gz",
+    sha256 = "<number>",
+)
+```
+
+Add the following to the project root *BUILD.bazel* file
+
+```py
+load("@io_k8s_rules_go//go:def.bzl", "code-generator")
+
+kubegen(
+    name = "kubegen",
+    cmd = "versioned",
+)
+```
+
+### Running the Bazel target
+
+```sh
+bazel run //:kubegen
```

Comment:
Are we going to check in the generated source? Are we going to regenerate automatically based on dependency changes?

Reply:
As long as we want to be go-get'able, we have to check in the code. kubegen will not change that.

Reply:
Neither supporting go get nor checking in code is discussed in this doc so the question is still open.

Reply:
It seems like kubegen should support both use cases, though I'd vote for checking in code since tools like glide and dep don't support generation.

If it's a goal to support not checking in the generated assetes, examples would be appreciated. cluster-registry tried this with the existing generators, and found it wasn't obvious how to structure the BUILD files so external projects could use them.

https://github.com/kubernetes/cluster-registry/issues/24#issuecomment-339485125

Reply:
Eventually, but not for the MVP.

Reply:
@mikedanese The Bazel build target is at the root directory and is mean to be invoked with run.  Why does it need to be prescriptive with whether consumers check the output into source control?

Reply:
having a separate `bazel run` command means that you can't just do `bazel build` and have it do the right thing.

for MVP this is fine, but you can certainly get more benefits by turning this into something more like a `genrule`.

Reply:
one other thing that might be worth noting is that calling `bazel run kubegen` may need to be followed by `bazel run gazelle` or whatever, since source files may change.